### DISCUSSION TITLE
Draft content available to anonymous users via SDK / GraphQL

### DISCUSSION THREAD

**Discussion by badrange** [directus/directus#12074] [Archive]
I have created a collection with a draft item in it, and would expect it not to be returned if I query for it while not being logged in and not if I access directus via the SDK without passing an admin users/editor user's token.

Is this a security bug or a directus feature?

Here is my GraphQL query:
```
{
	banners {
		location
		banners_translations (filter: { languages_code: { _eq: "fi" }}) {
			content
		}
		links {
			links_translations(filter: { languages_code: { _eq: "fi" } }) {
				title
				content
			}
		}
	}
}
```

**Comment by benhaynes** [comment]
The `draft` option within a `status` field is the same as any other data — there are no opinionated rules on what is or is not returned by the API... you have total control. The way this would typically be configured is to set the Role's permissions (for the API/SDK auth user) to filter out anything you don't want from READ. So you can say:

* Only if `status` equals `published`
* Only if `date_published` is less than or equal to now
* Only if `approved` equals `true`
* Etc...

This means that your SDK will only have access to what you deem "published"... which can be subjective.

CC @erondpowell ... this might be a good Guide/Article for the website

**Reply by badrange** [reply]
Thank you for the quick answer! I have to say this is confusing. Intuitively I'd expect draft content to only be available to authenticated users with at least some permission granted out of the box. 

Now I have a filter working according to your example.

I think a good default out of the box experience would be to have a filter for published for all new collections to the public role, it would be a "secure by default" option.
![image](https://user-images.githubusercontent.com/5602596/157692552-b86b419f-fce1-43ec-b013-3818a8ddfde7.png)

**Reply by benhaynes** [reply]
It just becomes very opinionated for us to add permissions automatically. By default ALL content is hidden for public and new roles... and admins have access to everything.

Keep in mind that these permissions control everything. You're talking about the role for an SDK "user" (presumably for data access to a public project), but removing draft content for actual user roles means they won't have access to that in the App either. Again, I thin the best way to handle this is to be _very_ safe by default (no permission to anything) and provide guides and shortcuts for admins to configure the permissions they want.

**Reply by badrange** [reply]
Now I realized that this was a classic PEBCAC: I had set the permissions at some point long ago and forgotten about it. So :heavy_plus_sign: :one:  to secure by default.

**Reply by badrange** [reply]
My next task is to figure out how to give permission to certain roles to see draft versions of content, while it is hidden for public users and non-editor users..

**Reply by benhaynes** [reply]
You just need some way to authorize users within your project code... then, you can use a different API access token with permission to fetch draft content. So normally, the project uses the public API which can't see draft... and then authorized users get access to draft content via the other API access token with elevated permissions.

The key is how you authorize those users. A query string, a full login page, a staging env, etc

**Reply by badrange** [reply]
Turns out I had already solved that :grin: I give every user who logs in to the frontend project their own token, and if they have the editor role they get to see the drafts, if they don't the draft items are hidden!

**Reply by rijkvanzanten** [reply]
I would also like to add that `draft` is just an arbitrary value in an arbitrary field 🙂 I wouldn't want to make any assumptions on whether or not you have a `status` field, and whether you consider `draft` to be visible/invisible by default 😄

**Reply by badrange** [reply]
Yes thank you for this clarification. As you might have guessed I have worked a lot with a highly opinionated CMS so I need to unlearn some habits / learn the mental models underpinning directus.
